include/ginkgo/core/matrix/csr.hpp

```diff
@@ -768,6 +768,74 @@ class Csr : public EnableLinOp<Csr<ValueType, IndexType>>,
 
     std::unique_ptr<LinOp> conj_transpose() const override;
 
+    /**
+     * Class describing the internal lookup structures created by
+     * multiply_reuse(const Csr*) to recompute a sparse matrix-matrix product
+     * with updated values.
+     */
+    class multiply_reuse_info {
+        friend class Csr;
+
+    public:
+        explicit multiply_reuse_info();
```

Comment:
private/delete?

Reply:
The purpose of making this constructor public is to mirror the behavior of the other reuse functions, allowing the reuse object to be declared outside of the loop where it is being used. Another option would be to ask people to put it into an `std::optional` or `std::unique_ptr`. I see your point, maybe we should talk briefly about the intended usage, e.g. currently
```cpp
multiply_reuse_info info;
std::unique_ptr<Csr> mtx;
while(...) {
    // ...
    if (!mtx) {
        std::tie(mtx, info) = A->multiply_reuse(B);
    } else {
        info->update_values(A, B, mtx);
    }
}
```
Can we do better than that?

Reply:
The alternative option might be
```cpp
multiply_reuse_info info;
std::unique_ptr<Csr> mtx;
while(...) {
    // ...
    A->multiply_reuse(B, mtx, info);
}
```

Reply:
The interface for permute_reuse with a default constructor is already part of 1.10, so we can't fix that there, but I'll deprecate that constructor and remove it here.